model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java

```diff
@@ -81,41 +87,66 @@ private void process(ArrayBlockingQueue<PersistentUpdate> queue) throws Interrup
             if (polled != null) {
                 batch.add(polled);
                 queue.drainTo(batch, maxBatchSize - 1);
-                try {
-                    LOG.debugf("Processing %d deferred session updates.", batch.size());
-                    Retry.executeWithBackoff(iteration -> {
-                                if (iteration < 2) {
-                                    // attempt to write whole batch in the first two attempts
-                                    KeycloakModelUtils.runJobInTransaction(factory,
-                                            innerSession -> batch.forEach(c -> c.perform(innerSession)));
-                                    batch.forEach(PersistentUpdate::complete);
-                                } else {
-                                    LOG.warnf("Running single changes in iteration %d for %d entries", iteration, batch.size());
-                                    ArrayList<PersistentUpdate> performedChanges = new ArrayList<>();
-                                    List<Throwable> throwables = new ArrayList<>();
-                                    batch.forEach(change -> {
-                                        try {
-                                            KeycloakModelUtils.runJobInTransaction(factory,
-                                                    change::perform);
-                                            change.complete();
-                                            performedChanges.add(change);
-                                        } catch (Throwable ex) {
-                                            throwables.add(ex);
+                KeycloakModelUtils.runJobInTransaction(factory, outerSession -> {
+                    TracingProvider tracing = outerSession.getProvider(TracingProvider.class);
+                    Tracer process = tracing.getTracer("PersistentSessionsWorker");
+                    SpanBuilder spanBuilder = process.spanBuilder("PersistentSessionsWorker.process");
+                    if (batch.size() > 1) {
+                        batch.stream().map(update -> update.getSpan().getSpanContext()).forEach(spanBuilder::addLink);
+                    } else {
+                        spanBuilder.setParent(Context.current().with(batch.get(0).getSpan()));
```

Comment:
This has to be consistent and if a single span is being generated by the batch it still needs to be a link... Assuming this is also an async processing.
There shouldn´t be a special case...

Reply:
resolved in next commit